syntax = "proto3";
package kvstore;

// Requests
message KeyValue {
  string key   = 1;
  string value = 2;
}
message Key {
  string key = 1;
}
message Prefix {
  string prefixKey = 1;
}
message Empty {}

// Responses
message GetResponse {
  string value = 1;
  string error = 2;
}
message GetPrefixResponse {
  repeated string values = 1;
}
message StatResponse {
  int32 total_sets        = 1;
  int32 total_gets        = 2;
  int32 total_getprefixes = 3;
}

// Servicio
service KVStore {
  rpc set       (KeyValue)       returns (Empty);
  rpc get       (Key)            returns (GetResponse);
  rpc getPrefix (Prefix)         returns (GetPrefixResponse);
  rpc stat      (Empty)          returns (StatResponse);
}